File: tyhp-lang/vscode/src/binary/BinaryManager.ts
import * as fs from "fs";
import * as vscode from "vscode";
import { InstallMode, normalizeReleaseTag, tagsMatch } from "../config/settingsCore";
import * as settings from "../config/settings";
import { Installer, fetchLatestRelease } from "./Installer";
import { deleteInstallMetadata, isExtensionOwnedInstall, readInstallMetadata } from "./metadata";
import { probeTyhpOnPath, validateTyhpPath } from "./PathProbe";
import { extensionInstallDir, isManagedInstallPath } from "./platform";
import { UpdateService } from "./UpdateService";

export type BinarySource = "setting" | "path" | "none";

/**
 * Result of locating the Tyhp CLI. Phase 4+ (LSP, tasks, XDebug proxy) must call
 * {@link resolveTyhpBinary} rather than reading `tyhp.path` on their own.
 */
export interface TyhpBinaryResolution {
    status: "ok" | "missing" | "invalid";
    /** Absolute filesystem path when `status` is `"ok"`. */
    executablePath?: string;
    message?: string;
    source: BinarySource;
    installMode: InstallMode;
}

const LAST_UPDATE_CHECK_KEY = "tyhp.binary.lastUpdateCheck";
const STARTUP_UPDATE_DELAY_MS = 5_000;
const UPDATE_CHECK_INTERVAL_MS = 6 * 60 * 60 * 1000;

let instance: BinaryManager | undefined;

export class BinaryManager implements vscode.Disposable {
    private readonly output: vscode.OutputChannel;
    private readonly resolutionEmitter = new vscode.EventEmitter<TyhpBinaryResolution>();
    private readonly installer: Installer;
    private readonly updates: UpdateService;
    private readonly cliDir: string;
    private resolution: TyhpBinaryResolution;
    private offeredFix = false;
    private readonly disposables: vscode.Disposable[] = [];

    readonly onDidChangeResolution = this.resolutionEmitter.event;

    constructor(private readonly context: vscode.ExtensionContext) {
        this.output = vscode.window.createOutputChannel("Tyhp");
        this.cliDir = extensionInstallDir(context.globalStorageUri.fsPath);
        this.installer = new Installer(context.globalStorageUri.fsPath, this.output);
        this.updates = new UpdateService(this.installer);
        this.resolution = missingResolution("Tyhp CLI has not been resolved yet.");
        this.disposables.push(
            this.output,
            this.resolutionEmitter,
            vscode.workspace.onDidChangeConfiguration((e) => {
                if (
                    e.affectsConfiguration("tyhp.path") ||
                    e.affectsConfiguration("tyhp.binary")
                ) {
                    void this.refreshResolution();
                }
            })
        );
    }

    get lastResolution(): TyhpBinaryResolution {
        return this.resolution;
    }

    async initialize(): Promise<void> {
        try {
            await this.probeAndPopulatePath();
            await this.refreshResolution();
            this.scheduleAutoUpdate();
        } catch (err) {
            const message = errorMessage(err);
            this.output.appendLine(`Activation error: ${message}`);
            this.setResolution(missingResolution(message));
        }
    }

    /**
     * Re-run PATH discovery when `tyhp.path` is empty, then validate. Never throws
     * (see {@link refreshResolution}); the "Refresh Tyhp binary" command surfaces
     * failures via the status bar instead of an unhandled rejection.
     */
    async refresh(): Promise<TyhpBinaryResolution> {
        try {
            await this.probeAndPopulatePath();
        } catch (err) {
            const message = errorMessage(err);
            this.output.appendLine(`PATH probe error: ${message}`);
        }
        return this.refreshResolution();
    }

    async resolve(): Promise<TyhpBinaryResolution> {
        return this.refreshResolution();
    }

    async installInteractive(): Promise<void> {
        const mode = await vscode.window.showQuickPick(
            [
                {
                    label: "Global",
                    description: "Install into a user-global location",
                    detail: "Never auto-updated. Updates only when you run this command. Sets `tyhp.path`.",
                    mode: "global" as const,
                },
                {
                    label: "Extension only",
                    description: "Install under this extension’s storage",
                    detail: "Auto-update and `tyhp.binary.pinnedVersion` apply. Sets `tyhp.path` to that binary.",
                    mode: "extension" as const,
                },
            ],
            {
                title: "Tyhp: Install / Update CLI",
                placeHolder: "Choose where to install the Tyhp compiler CLI",
                ignoreFocusOut: true,
            }
        );
        if (!mode) {
            return;
        }

        try {
            await vscode.window.withProgress(
                {
                    location: vscode.ProgressLocation.Notification,
                    title: "Installing Tyhp CLI",
                    cancellable: false,
                },
                async () => {
                    const pin = settings.getPinnedVersion();
                    const tag = pin !== "" ? normalizeReleaseTag(pin) : (await fetchLatestRelease()).tag_name;
                    this.output.appendLine(`Install / Update (${mode.mode}) → ${tag}`);
                    const result = await this.installer.install(mode.mode, tag);
                    await settings.setTyhpPath(result.executablePath);
                    await settings.setInstallMode(mode.mode);
                    if (mode.mode === "global") {
                        deleteInstallMetadata(this.cliDir);
                    }
                    await this.refreshResolution();
                    const pathNote =
                        mode.mode === "global"
                            ? ` If \`tyhp\` is not on PATH, add the install directory or keep using \`tyhp.path\`.`
                            : "";
                    void vscode.window.showInformationMessage(
                        `Installed Tyhp CLI ${result.version} at ${result.executablePath}.${pathNote}`
                    );
                }
            );
        } catch (err) {
            const message = errorMessage(err);
            this.output.appendLine(`Install failed: ${message}`);
            const pick = await vscode.window.showErrorMessage(`Tyhp CLI install failed: ${message}`, "Show Output");
            if (pick === "Show Output") {
                this.output.show(true);
            }
        }
    }

    async reveal(): Promise<void> {
        const resolved = this.resolution;
        if (resolved.status === "ok" && resolved.executablePath) {
            void vscode.window.showInformationMessage(`Tyhp CLI: ${resolved.executablePath}`);
            return;
        }
        const pick = await vscode.window.showErrorMessage(
            resolved.message ?? "Tyhp CLI is not available.",
            "Install / Update CLI"
        );
        if (pick === "Install / Update CLI") {
            await this.installInteractive();
        }
    }

    private async probeAndPopulatePath(): Promise<void> {
        if (!settings.tyhpPathIsUnset()) {
            return;
        }
        const found = probeTyhpOnPath();
        if (!found) {
            return;
        }
        let absolute = found;
        try {
            absolute = fs.realpathSync(found);
        } catch {
            absolute = found;
        }
        this.output.appendLine(`PATH probe found tyhp at ${absolute}; writing tyhp.path`);
        await settings.setTyhpPath(absolute);
        await settings.setInstallMode("path");
        deleteInstallMetadata(this.cliDir);
    }

    /**
     * Never throws: {@link resolveTyhpBinary} is the single resolve API later phases
     * (LSP, tasks, XDebug proxy) call, and the `tyhp.path` / `tyhp.binary.*` config-change
     * listener also awaits this directly. An unsupported platform, a transient fs error,
     * or any other unexpected failure must degrade to a `"missing"` resolution + status bar
     * error, not an unhandled rejection that later phases (or VS Code itself) would have to
     * guard against individually.
     */
    private async refreshResolution(): Promise<TyhpBinaryResolution> {
        try {
            return await this.doRefreshResolution();
        } catch (err) {
            const message = errorMessage(err);
            this.output.appendLine(`Resolution error: ${message}`);
            this.setResolution(missingResolution(message));
            await this.offerFix(message);
            return this.resolution;
        }
    }

    private async doRefreshResolution(): Promise<TyhpBinaryResolution> {
        const configured = settings.getTyhpPath();
        const installMode = settings.getInstallMode();

        if (configured !== "") {
            const check = validateTyhpPath(configured);
            if (check.ok && check.absolutePath) {
                this.setResolution({
                    status: "ok",
                    executablePath: check.absolutePath,
                    source: "setting",
                    installMode,
                });
                return this.resolution;
            }
            const message =
                check.message ??
                `Tyhp CLI at \`${configured}\` is missing or is not a file. Use “Tyhp: Install / Update CLI” or fix \`tyhp.path\`.`;
            this.setResolution({
                status: "invalid",
                executablePath: check.absolutePath,
                message,
                source: "setting",
                installMode,
            });
            await this.offerFix(message);
            return this.resolution;
        }

        const probed = probeTyhpOnPath();
        if (probed) {
            this.setResolution({
                status: "ok",
                executablePath: probed,
                source: "path",
                installMode,
            });
            return this.resolution;
        }

        const message =
            "Tyhp CLI was not found. Use “Tyhp: Install / Update CLI” or set `tyhp.path` to an existing binary.";
        this.setResolution(missingResolution(message, installMode));
        await this.offerFix(message);
        return this.resolution;
    }

    private async offerFix(message: string): Promise<void> {
        if (this.offeredFix) {
            return;
        }
        this.offeredFix = true;
        const pick = await vscode.window.showErrorMessage(message, "Install / Update CLI", "Show Output");
        if (pick === "Install / Update CLI") {
            await this.installInteractive();
        } else if (pick === "Show Output") {
            this.output.show(true);
        }
    }

    private scheduleAutoUpdate(): void {
        const handle = setTimeout(() => {
            void this.runAutoUpdate();
        }, STARTUP_UPDATE_DELAY_MS);
        this.disposables.push({ dispose: () => clearTimeout(handle) });
    }

    private async runAutoUpdate(): Promise<void> {
        const installMode = settings.getInstallMode();
        const metadata = readInstallMetadata(this.cliDir);

        // `tyhp.binary.installMode` can drift from `tyhp.path` when a user hand-edits
        // settings.json (e.g. points `tyhp.path` at a custom build) without going through
        // "Tyhp: Install / Update CLI" or PATH re-probe, either of which would reset
        // `installMode`/metadata. Guard against auto-updating (and overwriting `tyhp.path`
        // with) the extension-managed binary in that case: "setting wins" must hold even
        // when the stale `installMode` still says `extension`.
        if (
            installMode === "extension" &&
            !isManagedInstallPath(settings.getTyhpPath(), this.context.globalStorageUri.fsPath)
        ) {
            this.output.appendLine(
                "Skipping auto-update: tyhp.path no longer points at the extension-managed install."
            );
            return;
        }

        const pin = settings.getPinnedVersion();
        const pinNeedsApply =
            pin !== "" &&
            installMode === "extension" &&
            isExtensionOwnedInstall(metadata) &&
            !tagsMatch(metadata?.version ?? "", pin);

        if (!pinNeedsApply) {
            const last = this.context.globalState.get<number>(LAST_UPDATE_CHECK_KEY, 0);
            if (Date.now() - last < UPDATE_CHECK_INTERVAL_MS) {
                this.output.appendLine("Skipping auto-update check (debounced).");
                return;
            }
        }

        try {
            const result = await this.updates.checkAndApply(
                {
                    installMode,
                    autoUpdate: settings.getAutoUpdate(),
                    pinnedVersion: pin,
                },
                metadata,
                this.output
            );
            await this.context.globalState.update(LAST_UPDATE_CHECK_KEY, Date.now());
            if (result.updated && result.path) {
                await settings.setTyhpPath(result.path);
                await this.refreshResolution();
                void vscode.window.showInformationMessage(`Updated Tyhp CLI (${result.reason}).`);
            }
        } catch (err) {
            const message = errorMessage(err);
            this.output.appendLine(`Auto-update failed: ${message}`);
        }
    }

    private setResolution(resolution: TyhpBinaryResolution): void {
        this.resolution = resolution;
        this.resolutionEmitter.fire(resolution);
    }

    dispose(): void {
        for (const d of this.disposables) {
            d.dispose();
        }
        this.disposables.length = 0;
    }
}

function missingResolution(message: string, installMode: InstallMode = settings.getInstallMode()): TyhpBinaryResolution {
    return {
        status: "missing",
        message,
        source: "none",
        installMode,
    };
}

function errorMessage(err: unknown): string {
    return err instanceof Error ? err.message : String(err);
}

export function registerBinaryManager(context: vscode.ExtensionContext): BinaryManager {
    instance = new BinaryManager(context);
    context.subscriptions.push(instance);
    return instance;
}

/**
 * Resolve the Tyhp CLI executable. Later phases (language server, tasks,
 * XDebug proxy) should call this instead of reading `tyhp.path` directly.
 */
export async function resolveTyhpBinary(): Promise<TyhpBinaryResolution> {
    if (!instance) {
        return {
            status: "missing",
            source: "none",
            installMode: "path",
            message: "Tyhp binary manager is not active.",
        };
    }
    return instance.resolve();
}

/** Last cached resolution without hitting the filesystem again. */
export function getResolvedTyhpBinary(): TyhpBinaryResolution {
    if (!instance) {
        return {
            status: "missing",
            source: "none",
            installMode: "path",
            message: "Tyhp binary manager is not active.",
        };
    }
    return instance.lastResolution;
}

export function getBinaryManager(): BinaryManager | undefined {
    return instance;
}

File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/binary/BinaryManager.kt
package com.tyhp.lang.binary

import com.intellij.notification.NotificationAction
import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.PathManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.options.ShowSettingsUtil
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.util.Disposer
import com.tyhp.lang.settings.InstallMode
import com.tyhp.lang.settings.TyhpConfigurable
import com.tyhp.lang.settings.TyhpSettings
import com.tyhp.lang.settings.normalizeReleaseTag
import com.tyhp.lang.settings.tagsMatch
import java.nio.file.Path
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean

private const val STARTUP_UPDATE_DELAY_MS = 5_000L
private const val UPDATE_CHECK_INTERVAL_MS = 6L * 60L * 60L * 1000L
private const val NOTIFICATION_GROUP = "Tyhp Language"

@Service(Service.Level.APP)
class BinaryManager {
    private val log = thisLogger()
    private val started = AtomicBoolean(false)
    private val offeredFix = AtomicBoolean(false)
    private val autoUpdateScheduled = AtomicBoolean(false)

    @Volatile
    var lastResolution: TyhpBinaryResolution = inactiveResolution()
        private set

    private val resolutionListeners = CopyOnWriteArrayList<BinaryResolutionListener>()

    /**
     * Subscribe to CLI resolution changes (PATH probe, settings, install).
     * Later phases (LSP) restart processes from this instead of reading `tyhp.path`.
     */
    fun addResolutionListener(parentDisposable: Disposable, listener: BinaryResolutionListener) {
        resolutionListeners.add(listener)
        Disposer.register(parentDisposable) { resolutionListeners.remove(listener) }
    }

    private fun publishResolution(project: Project?, resolution: TyhpBinaryResolution): TyhpBinaryResolution {
        val previous = lastResolution
        lastResolution = resolution
        val unchanged =
            previous.status == resolution.status &&
                previous.executablePath == resolution.executablePath &&
                previous.message == resolution.message
        if (unchanged) {
            return resolution
        }
        for (listener in resolutionListeners) {
            try {
                listener.onResolutionChanged(project, resolution)
            } catch (err: Throwable) {
                log.warn("Binary resolution listener failed: ${errorMessage(err)}", err)
            }
        }
        return resolution
    }

    private val pluginStoragePath: String
        get() = Path.of(PathManager.getConfigPath(), "tyhp-lang").toString()

    private val cliDir: Path
        get() = pluginInstallDir(pluginStoragePath)

    fun onProjectOpened(project: Project) {
        try {
            // Probe on every project open: this service is app-scoped, but each
            // project can have its own unset `tyhp.path` project override, so a
            // single app-wide latch here would leave later-opened projects (in a
            // multi-window PhpStorm session) unprobed. probeAndPopulatePath is a
            // no-op once a path is set, so re-running it per project is cheap.
            probeAndPopulatePath(project)
            if (started.compareAndSet(false, true)) {
                scheduleAutoUpdate()
            }
            refreshResolution(project)
        } catch (err: Throwable) {
            val message = errorMessage(err)
            log.warn("Tyhp binary activation error: $message", err)
            publishResolution(project, missingResolution(message))
            notifyError(project, message)
        }
    }

    /**
     * Re-run PATH discovery when `tyhp.path` is empty, then validate. Never throws.
     */
    fun refresh(project: Project?): TyhpBinaryResolution {
        try {
            probeAndPopulatePath(project)
        } catch (err: Throwable) {
            val message = errorMessage(err)
            log.warn("PATH probe error: $message", err)
        }
        return refreshResolution(project)
    }

    fun resolve(project: Project?): TyhpBinaryResolution = refreshResolution(project)

    fun installInteractive(project: Project?) {
        val choice = Messages.showDialog(
            project,
            "Choose where to install the Tyhp compiler CLI.\n\n" +
                "Global installs into ~/.local/bin/tyhp (Unix) or %LOCALAPPDATA%\\Programs\\tyhp\\tyhp.exe (Windows) " +
                "and is never auto-updated.\n\n" +
                "Plugin only installs under this plugin’s config storage. Auto-update and tyhp.binary.pinnedVersion apply.",
            "Tyhp: Install / Update CLI",
            arrayOf("Global", "Plugin only", "Cancel"),
            0,
            Messages.getQuestionIcon(),
        )
        val mode = when (choice) {
            0 -> "global"
            1 -> "extension"
            else -> return
        }
        runInstall(project, mode)
    }

    fun reveal(project: Project?) {
        val resolved = lastResolution
        if (resolved.status == BinaryStatus.OK && !resolved.executablePath.isNullOrBlank()) {
            Messages.showInfoMessage(project, "Tyhp CLI: ${resolved.executablePath}", "Tyhp CLI Path")
            return
        }
        val message = resolved.message ?: "Tyhp CLI is not available."
        val pick = Messages.showDialog(
            project,
            message,
            "Tyhp CLI",
            arrayOf("Install / Update CLI", "Open Settings", "Cancel"),
            0,
            Messages.getErrorIcon(),
        )
        when (pick) {
            0 -> installInteractive(project)
            1 -> openSettings(project)
        }
    }

    private fun runInstall(project: Project?, mode: String) {
        ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Installing Tyhp CLI", false) {
            override fun run(indicator: ProgressIndicator) {
                indicator.isIndeterminate = true
                try {
                    val installer = Installer(pluginStoragePath, InstallerLogger { line -> log.info(line) })
                    val pin = TyhpSettings.getPinnedVersion()
                    val tag = if (pin.isNotEmpty()) {
                        normalizeReleaseTag(pin)
                    } else {
                        fetchLatestRelease().tagName
                    }
                    log.info("Install / Update ($mode) → $tag")
                    val result = installer.install(mode, tag)
                    ApplicationManager.getApplication().invokeLater {
                        TyhpSettings.setTyhpPath(project, result.executablePath)
                        TyhpSettings.setInstallMode(if (mode == "extension") InstallMode.EXTENSION else InstallMode.GLOBAL)
                        if (mode == "global") {
                            deleteInstallMetadata(cliDir)
                        }
                        refreshResolution(project)
                        val pathNote = if (mode == "global") {
                            " If `tyhp` is not on PATH, add the install directory or keep using `tyhp.path`."
                        } else {
                            ""
                        }
                        notifyInfo(
                            project,
                            "Installed Tyhp CLI ${result.version} at ${result.executablePath}.$pathNote",
                        )
                    }
                } catch (err: Throwable) {
                    val message = errorMessage(err)
                    log.warn("Install failed: $message", err)
                    ApplicationManager.getApplication().invokeLater {
                        publishResolution(project, missingResolution(message, TyhpSettings.getInstallMode()))
                        notifyError(project, "Tyhp CLI install failed: $message")
                    }
                }
            }
        })
    }

    private fun probeAndPopulatePath(project: Project?) {
        if (!TyhpSettings.tyhpPathIsUnset(project)) {
            return
        }
        val found = probeTyhpOnPath() ?: return
        var absolute = found
        try {
            absolute = Path.of(found).toRealPath().toString()
        } catch (_: Exception) {
            absolute = found
        }
        log.info("PATH probe found tyhp at $absolute; writing tyhp.path")
        TyhpSettings.setTyhpPath(project, absolute)
        TyhpSettings.setInstallMode(InstallMode.PATH)
        deleteInstallMetadata(cliDir)
    }

    /**
     * Never throws: [resolveTyhpBinary] is the single resolve API later phases call.
     */
    private fun refreshResolution(project: Project?): TyhpBinaryResolution {
        return try {
            doRefreshResolution(project)
        } catch (err: Throwable) {
            val message = errorMessage(err)
            log.warn("Resolution error: $message", err)
            offerFix(project, message)
            return publishResolution(project, missingResolution(message))
        }
    }

    private fun doRefreshResolution(project: Project?): TyhpBinaryResolution {
        val configured = TyhpSettings.getTyhpPath(project)
        val installMode = TyhpSettings.getInstallMode()

        if (configured.isNotEmpty()) {
            val check = validateTyhpPath(configured)
            if (check.ok && !check.absolutePath.isNullOrBlank()) {
                return publishResolution(
                    project,
                    TyhpBinaryResolution(
                        status = BinaryStatus.OK,
                        executablePath = check.absolutePath,
                        source = BinarySource.SETTING,
                        installMode = installMode,
                    ),
                )
            }
            val message = check.message
                ?: "Tyhp CLI at `$configured` is missing or is not a file. Use “Tyhp: Install / Update CLI” or fix `tyhp.path`."
            offerFix(project, message)
            return publishResolution(
                project,
                TyhpBinaryResolution(
                    status = BinaryStatus.INVALID,
                    executablePath = check.absolutePath,
                    message = message,
                    source = BinarySource.SETTING,
                    installMode = installMode,
                ),
            )
        }

        val probed = probeTyhpOnPath()
        if (probed != null) {
            return publishResolution(
                project,
                TyhpBinaryResolution(
                    status = BinaryStatus.OK,
                    executablePath = probed,
                    source = BinarySource.PATH,
                    installMode = installMode,
                ),
            )
        }

        val message =
            "Tyhp CLI was not found. Use “Tyhp: Install / Update CLI” or set `tyhp.path` to an existing binary."
        offerFix(project, message)
        return publishResolution(project, missingResolution(message, installMode))
    }

    private fun offerFix(project: Project?, message: String) {
        if (!offeredFix.compareAndSet(false, true)) {
            return
        }
        notifyError(project, message)
    }

    private fun scheduleAutoUpdate() {
        if (!autoUpdateScheduled.compareAndSet(false, true)) {
            return
        }
        val app = ApplicationManager.getApplication()
        app.executeOnPooledThread {
            try {
                Thread.sleep(STARTUP_UPDATE_DELAY_MS)
            } catch (_: InterruptedException) {
                return@executeOnPooledThread
            }
            if (app.isDisposed) {
                return@executeOnPooledThread
            }
            runAutoUpdate()
        }
    }

    private fun runAutoUpdate() {
        val installMode = TyhpSettings.getInstallMode()
        val configuredPath = TyhpSettings.getTyhpPath(null)
        val platform = try {
            detectHostPlatform()
        } catch (err: UnsupportedPlatformError) {
            log.info(err.message)
            return
        }

        if (shouldSkipAutoUpdateDueToPathDrift(installMode, configuredPath, pluginStoragePath, platform)) {
            log.info("Skipping auto-update: tyhp.path no longer points at the plugin-managed install.")
            return
        }

        val metadata = readInstallMetadata(cliDir)
        val pin = TyhpSettings.getPinnedVersion()
        val pinNeedsApply =
            pin.isNotEmpty() &&
                installMode == InstallMode.EXTENSION &&
                isPluginOwnedInstall(metadata) &&
                !tagsMatch(metadata?.version ?: "", pin)

        if (!pinNeedsApply) {
            val last = TyhpSettings.getLastUpdateCheckEpochMs()
            if (System.currentTimeMillis() - last < UPDATE_CHECK_INTERVAL_MS) {
                log.info("Skipping auto-update check (debounced).")
                return
            }
        }

        try {
            val installer = Installer(pluginStoragePath, InstallerLogger { line -> log.info(line) }, platform)
            val result = UpdateService(installer).checkAndApply(
                UpdateServiceSettings(
                    installMode = installMode,
                    autoUpdate = TyhpSettings.getAutoUpdate(),
                    pinnedVersion = pin,
                ),
                metadata,
                InstallerLogger { line -> log.info(line) },
            )
            TyhpSettings.setLastUpdateCheckEpochMs(System.currentTimeMillis())
            if (result.updated && result.path != null) {
                TyhpSettings.setTyhpPath(null, result.path)
                refreshResolution(null)
                notifyInfo(null, "Updated Tyhp CLI (${result.reason}).")
            }
        } catch (err: Throwable) {
            val message = errorMessage(err)
            log.warn("Auto-update failed: $message", err)
        }
    }

    private fun notifyError(project: Project?, message: String) {
        val notification = NotificationGroupManager.getInstance()
            .getNotificationGroup(NOTIFICATION_GROUP)
            .createNotification("Tyhp CLI", message, NotificationType.ERROR)
        notification.addAction(
            NotificationAction.createSimple("Install / Update CLI") {
                installInteractive(project)
            },
        )
        notification.addAction(
            NotificationAction.createSimple("Open Settings") {
                openSettings(project)
            },
        )
        notification.notify(project)
    }

    private fun notifyInfo(project: Project?, message: String) {
        NotificationGroupManager.getInstance()
            .getNotificationGroup(NOTIFICATION_GROUP)
            .createNotification("Tyhp CLI", message, NotificationType.INFORMATION)
            .notify(project)
    }

    private fun openSettings(project: Project?) {
        if (project != null && !project.isDisposed) {
            ShowSettingsUtil.getInstance().showSettingsDialog(project, TyhpConfigurable::class.java)
        }
    }

    companion object {
        fun getInstance(): BinaryManager =
            ApplicationManager.getApplication().getService(BinaryManager::class.java)
    }
}

/**
 * Resolve the Tyhp CLI executable. Later phases (language server, run configs,
 * XDebug proxy) should call this instead of reading `tyhp.path` directly.
 */
fun resolveTyhpBinary(project: Project? = null): TyhpBinaryResolution {
    val app = ApplicationManager.getApplication() ?: return inactiveResolution()
    val manager = app.getService(BinaryManager::class.java) ?: return inactiveResolution()
    return manager.resolve(project)
}

/** Last cached resolution without hitting the filesystem again. */
fun getResolvedTyhpBinary(): TyhpBinaryResolution {
    val app = ApplicationManager.getApplication() ?: return inactiveResolution()
    val manager = app.getService(BinaryManager::class.java) ?: return inactiveResolution()
    return manager.lastResolution
}

internal fun errorMessage(err: Throwable): String = err.message ?: err.toString()
